pkg/tfbridge/validate_input_types.go

```diff
@@ -0,0 +1,361 @@
+		//
+		// Specifically it defines the defaultType, discriminator, mapping and elements.
+		if len(typeSpec.OneOf) < 2 {
+			// invalid type, don't validate
+			return nil
+		}
+
+		// don't validate types with Discriminator (yet)
+		if typeSpec.Discriminator != nil {
+			return nil
+		}
+
+		// default type
+		if typeSpec.Type != "" {
+			typeSpec.OneOf = append(typeSpec.OneOf, pschema.TypeSpec{Type: typeSpec.Type})
+		}
+
+		failures := []TypeFailure{}
+		for _, spec := range typeSpec.OneOf {
+			failure := v.validatePropertyValue(propertyValue, spec, propertyPath)
+			if len(failure) == 0 {
+				// we have found the correct type
+				return nil
+			}
+			failures = append(failures, failure...)
+		}
+		return failures
+	}
+
+	switch typeSpec.Type {
+	case "bool":
+		// The bridge permits coalescing strings to booleans, hence skip strings.
+		if !propertyValue.IsBool() && !propertyValue.IsString() {
+			return []TypeFailure{{
+				ResourcePath: propertyPath.String(),
+				Reason: fmt.Sprintf(
+					"expected boolean type, got %s type",
+					propertyValue.TypeString(),
+				),
+			}}
+		}
+		return nil
+	case "integer", "number":
+		// The bridge permits coalescing strings to numbers, hence skip strings.
+		if !propertyValue.IsNumber() && !propertyValue.IsString() {
+			return []TypeFailure{{
+				ResourcePath: propertyPath.String(),
+				Reason: fmt.Sprintf(
+					"expected number type, got %s type",
+					propertyValue.TypeString(),
+				),
+			}}
+		}
+		return nil
+	case "string":
+		// The bridge permits coalescing numbers and booleans to strings, hence skip these.
+		if !propertyValue.IsString() && !propertyValue.IsNumber() && !propertyValue.IsBool() {
+			return []TypeFailure{{
+				ResourcePath: propertyPath.String(),
+				Reason: fmt.Sprintf(
+					"expected string type, got %s type",
+					propertyValue.TypeString(),
+				),
+			}}
+		}
+		return nil
+	case "array":
+		if propertyValue.IsArray() {
+			if typeSpec.Items == nil {
+				// Unknown item type so nothing more to check.
+				return nil
+			}
+			// Check every item against the array element type.
+			failures := []TypeFailure{}
+			for idx, arrayValue := range propertyValue.ArrayValue() {
+				pb := append(propertyPath, idx)
+				failure := v.validatePropertyValue(arrayValue, *typeSpec.Items, pb)
+				if failure != nil {
+					failures = append(failures, failure...)
+				}
+			}
+			return failures
+		}
+		return []TypeFailure{{
+			ResourcePath: propertyPath.String(),
+			Reason: fmt.Sprintf(
+				"expected array type, got %s type",
+				propertyValue.TypeString(),
+			),
+		}}
+	case "object":
+		// This is not really an object but a map type with some element type, which is assumed to be string if
+		// unspecified. Check accordingly. This should be very similar to the "array" case.
+		//
+		// elementTypeSpec := typeSpec.AdditionalProperties
+		if propertyValue.IsObject() {
+			if typeSpec.AdditionalProperties == nil {
+				// Unknown item type so nothing more to check
```

Comment:
The binder actually assumes this would be a stringly typed map[string]string. Could go either way here.

Reply:
Are you saying that if `Type: "object", AdditionalProperties: nil` then the assumption is that the type is `Type: "object", AdditionalProperties: { Type: "string" }`?

Reply:
Yes! Last I checked that seemed to be literally what the binder code was saying.  It is not easy to tell though if that's intentional or "let's assume some type here". I guess I prefer your version here for now - no type specified so we won't check.